Nft tracker

KarishmaBothara opened the pull request:
Description - 
This is to track all the nft related data
Set up
1. run docker-compose up to start mongo db service, 
2. run yarn nft:stats -- services scans the block and pushes data in db
3. run yarn api --  api services which provides the required historic data
    a. http://localhost:3000/nft/token/56/35/0 - end point to get history of an nft with collection id 56, series id 35 and serialNumber 0
    b. http://localhost:3000/nft/listing/653 - get all information history for a listing id
    c. http://localhost:3000/nft/wallet/5D2XVbob7zYjWzhkZ8nPcBgR7yEhAiXE5rYV8qhKdktrcHSq - get all the nft transaction with a wallet

KarishmaBothara committed:
Litho stats

KarishmaBothara committed:
NFT event tracker

KarishmaBothara committed:
remove commented code

KarishmaBothara committed:
store formatted balance in db

ken-futureverse commented on src/api/controller/nfts.ts:
```diff
@@ -0,0 +1,158 @@
+import { nftsModel } from "../models/nftsModel";
+
+export async function getTokenDetails(request, reply) {
```
can we type annotate function parameters and the return plz.

ken-futureverse commented on src/api/controller/nfts.ts:
```diff
@@ -0,0 +1,158 @@
+	const { collectionId, seriesId, serialNumber } = request.params;
+	const tokenId = `[${collectionId},${seriesId},${serialNumber}]`;
+	const data = await nftsModel.fetchEventStream(tokenId);
+	if (data[0]) {
```
probably best to check for length plus use optional chaining to avoid unexpected `data` plus can flatten this `if` statement by return immediate if `data[0]` doesn't exist, so

```js
...
if(!data?.length) return reply.status(500).send({ error: "Token Not found!" });
...

return reply.status(200).send(response);
```

ken-futureverse commented on src/api/controller/nfts.ts:
```diff
@@ -0,0 +1,158 @@
+		// Listing can happen multiple times
+		// find the recent/last listing id for a token
+		const listingStartedDetails = [...nftData]
+			.reverse()
```
by `reverse()` you just want it to have a good chance of finding it faster?

KarishmaBothara replied:
a token can be listed multiple time.. I would be interested in showing the latest  listing id at the upper layer and all the history will follow

ken-futureverse commented on src/api/controller/nfts.ts:
```diff
@@ -0,0 +1,158 @@
+		const listingClosedDetails = [...nftData]
+			.reverse()
+			.find((nft) => nft.data.eventType === "LISTING_CLOSED");
+		if (listingStartedDetails) {
```
can we flatten this `if` block and looks like we don't do anything for the `true` part?

ken-futureverse commented on src/api/controller/nfts.ts:
```diff
@@ -0,0 +1,158 @@
+	const data = await nftsModel.fetchEventStream(request.params.listingId);
+	const listingData = data[0];
+	console.log("listingData:", listingData);
+	if (listingData.length > 0) {
```
same as above, use optional chaining `data?.[0]?.length` plus return early

ken-futureverse commented on src/api/controller/nfts.ts:
```diff
@@ -0,0 +1,158 @@
+	return reply.status(500).send({ error: "Token Not found!" });
+}
+
+export async function getListingDetails(request, reply) {
```
type annotate for function parameters and return please

ken-futureverse commented on src/api/controller/nfts.ts:
```diff
@@ -0,0 +1,158 @@
+	}
+}
+
+export async function getWalletDetails(request, reply) {
```
similar feedbacks as above, type annotation, optional chaining and return early :)

ken-futureverse commented on src/api/index.ts:
```diff
@@ -0,0 +1,18 @@
+import Fastify from "fastify";
```
seems like a well-designed package, should we pick it as our `defactor` standard when writing Web API?

ken-futureverse commented on src/api/models/nftsModel.ts:
```diff
@@ -0,0 +1,36 @@
+		let res = [{}];
+		try {
+			res = await connection.execute(
+				`SELECT * FROM event_tracker where stream_id = ? ORDER BY version ASC`,
```
worth to consider using an ORM like https://www.prisma.io/docs/ to help you interface with DB, safer

KarishmaBothara replied:
will check

KarishmaBothara replied:
done

ken-futureverse commented on src/api/models/nftsModel.ts:
```diff
@@ -0,0 +1,36 @@
+import { DATABASE } from "../config/database";
+
+export const nftsModel = {
+	fetchEventStream: async function (tokenId) {
```
you can export them individually, rather have it grouped like this

ken-futureverse commented on src/api/routes/index.ts:
```diff
@@ -0,0 +1,14 @@
+	);
+	fastify.get("/nft/listing/:listingId", getListingDetails);
+	fastify.get("/nft/wallet/:address", getWalletDetails);
+}
```
❤️  `fastify`

ken-futureverse commented on src/scanner/scanNFTStats.ts:
```diff
@@ -0,0 +1,364 @@
+	const eventData = findNFTEvent ? findNFTEvent.event.data.toJSON() : null;
+	switch (method) {
+		case "mintUnique": {
+			if (eventData) {
```
flatten and return early please

ken-futureverse commented on src/scanner/utility.ts:
```diff
@@ -0,0 +1,1031 @@
+let blockTime;
+
+// Convert a block number to date
+async function convertBlockToDate(api, blockNumber, date) {
```
move this to its own function

ken-futureverse commented on src/scanner/utility.ts:
```diff
@@ -0,0 +1,1031 @@
+			owner
+		);
+	} catch (e) {
+		console.log(
```
I think we should re throw the error as this point, so the nftsModel can catch it all and handle them accordingly

ken-futureverse commented on src/scanner/utility.ts:
```diff
@@ -0,0 +1,1031 @@
+						seriesId
+					)
+			  ).toHuman();
+		const tokenData = {
```
instead of await one after another, if they can happen in parallel then you can utilise the `Promise.all`, that will shave off a few `ms`

```
const [nextSerialNumber, imgUrl] = await Promise.all([
await api.query.nft.nextSerialNumber.at(blockHash, collectionId, seriesId),
api.query.nft.seriesMetadataScheme
			? (
					await api.query.nft.seriesMetadataScheme.at(
						blockHash,
						collectionId,
						seriesId
					)
			  ).toHuman()
			: (
					await api.query.nft.seriesMetadataURI.at(
						blockHash,
						collectionId,
						seriesId
					)
			  ).toHuman()
])
```

I might have not picked up everything, so please use this pattern anywhere if it is applicable. Will help speed up script

KarishmaBothara replied:
sure..

ken-futureverse commented on src/scanner/utility.ts:
```diff
@@ -0,0 +1,1031 @@
+				owner: owner,
+				txHash: txHash,
+			},
+			eventType: "SERIES NAMED",
```
should be `SERIES_NAMED`

ken-futureverse commented on src/scanner/utility.ts:
```diff
@@ -0,0 +1,1031 @@
+				owner: newOwner,
+				txHash: txHash,
+			},
+			eventType: "TRANSFER",
```
is it to track transfer of NFT? then should it be `NFT_TRANSFERED`?

ken-futureverse commented on src/scanner/utility.ts:
```diff
@@ -0,0 +1,1031 @@
+				owner: newOwner,
+				txHash: txHash,
+			},
+			eventType: "TRANSFER",
```
similar to the NFT_TRANSFERED comment above

ken-futureverse commented on src/scanner/utility.ts:
```diff
@@ -0,0 +1,1031 @@
+				owner: null,
+				txHash: txHash,
+			},
+			eventType: "BURN",
```
`NFT_BURNED`

ken-futureverse commented on src/scanner/utility.ts:
```diff
@@ -0,0 +1,1031 @@
+				owner: null,
+				txHash: txHash,
+			},
+			eventType: "BURN",
```
ditto

ken-futureverse reviewed (changes requested):
Seems fine, I have a few comments, the main things are
- [x] Break the `utility.ts` into smaller files, ideally one function each
- [x] Annotate type for function parameters and return
- [x] Ensure `eventType` naming conventions are consistent
- [x] Flatten `try ... catch`, `if` statements to make code easier to read
- [x] Utilise `Promise.all` anywhere to speed up the processing script

KarishmaBothara committed:
break utility into smaller files

KarishmaBothara committed:
make event type consistent everywhere

KarishmaBothara committed:
as per review comment

KarishmaBothara committed:
as per review comment

KarishmaBothara committed:
use prisma client for db operations

KarishmaBothara commented:
addressed all the review comments.. need to think on how to handle error in case of db error for example duplicate entry error.. do we really want to re throw the error and stop the app.. as we can safely let these kind of error go.. by catching it and logging it

ken-futureverse reviewed (commented):
I see a mixture of `console.log` and `logger.`, is that intentional? Should we use one logging system

Re error, are there any *fatal* type error you could think off? If not a error logging out should be fine.

KarishmaBothara committed:
replace console log by logger

KarishmaBothara commented:
> I see a mixture of `console.log` and `logger.`, is that intentional? Should we use one logging system
> 
> Re error, are there any _fatal_ type error you could think off? If not a error logging out should be fine.

Updated to use logger everywhere.. No, nothing in particular I can think of in terms of error... But we can run this scanner at some place and monitor logs, then will be able to understand and update accordingly.

KarishmaBothara committed:
get api.At for older runtme version to find call

KarishmaBothara committed:
use mongo db

KarishmaBothara committed:
update

ken-futureverse commented:
this has been moved to https://github.com/cennznet/api-nft/pull/1 can we close this @KarishmaBothara ?
